app/src/main/java/com/github/featuredetectandroid/MainActivity.kt

```diff
@@ -60,62 +48,57 @@ class MainActivity : ComponentActivity() {
+
+        // SuperPoint, which needs a context during initialization,
+        // does not contain it inside itself, and therefore does not require any processing
+        // in the oDestroy and onResume.
+        outputViewModel.featureDetector = KeypointDetectionAlgorithm.nameToFeatureDetector(
+            algorithmName = preferencesManager.getSelectedAlgorithm(),
+            context = this,
+            width = outputViewModel.featureDetector?.width ?: 0,
+            height = outputViewModel.featureDetector?.height ?: 0
+        )
 
         setContent {
             FeatureDetectAppTheme {
                 var selectedAlgorithm by remember {
                     mutableStateOf(preferencesManager.getSelectedAlgorithm())
                 }
-                Scaffold(
-                    drawerContent = {
-                        Menu(
-                            header = "Keypoint detection algorithm:",
-                            options = KeypointDetectionAlgorithm.names,
-                            selectedOption = selectedAlgorithm,
-                            onSelected = { algorithmName ->
-                                preferencesManager.putSelectedAlgorithm(algorithmName)
-                                selectedAlgorithm = algorithmName
-                            }
-                        )
-                    }
-                ) {
-                    Box(Modifier.fillMaxSize()) {
-                        if (imageViewModel.isCameraPermissionGranted) {
-                            imageViewModel.grayscaleBitmap?.let { bitmap ->
-                                Image(
-                                    bitmap = bitmap.asImageBitmap(),
-                                    modifier = Modifier.fillMaxSize(),
-                                    contentDescription = "Grayscale photo"
-                                )
-                            }
-                        } else {
-                            Text(
-                                text = "Camera permission required",
-                                modifier = Modifier.align(Alignment.Center)
+                AppLayout(
+                    isCameraPermissionGranted = isCameraPermissionGranted(),
+                    keypointOffsets = outputViewModel.keypointOffsets,
+                    frameBitmap = outputViewModel.frameBitmap,
+                    selectedAlgorithm = selectedAlgorithm,
+                    onAlgorithmSelected = { algorithmName ->
+                        preferencesManager.putSelectedAlgorithm(algorithmName)
+                        selectedAlgorithm = algorithmName
+                        outputViewModel.keypointOffsets = emptyList()
+                        outputViewModel.featureDetector =
+                            KeypointDetectionAlgorithm.nameToFeatureDetector(
+                                algorithmName = preferencesManager.getSelectedAlgorithm(),
```

Comment:
Let's use `algorithmName` here -- should be quicker

Reply:
fixed